kernel/runqueue_realtime/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+/// If the task is aperiodic, `period` will contain the value `None`
+/// `RealtimeTaskRef` implements `Deref` and `DerefMut` traits, which dereferences to `TaskRef`
+#[derive(Debug, Clone)]
+pub struct RealtimeTaskRef {
+    /// `TaskRef` wrapped by `RealtimeTaskRef`
+    taskref: TaskRef,
+
+    /// If the task is periodic, this value will be the period in ticks wrapped in `Some`
+    /// If the task is aperiodic, this value will be `None`
+    period: Option<usize>,
+
+    /// Number of context switches the task has undergone. Not used in scheduling algorithm
+    context_switches: usize,
+}
+
+impl Deref for RealtimeTaskRef {
+    type Target = TaskRef;
+    fn deref(&self) -> &TaskRef {
+        &self.taskref
+    }
+}
+
+impl DerefMut for RealtimeTaskRef {
+    fn deref_mut(&mut self) -> &mut TaskRef {
+        &mut self.taskref
+    }
+}
+
+impl RealtimeTaskRef {
+    /// Creates a new `RealtimeTaskRef` that wraps the given `TaskRef`
+    pub fn new(taskref: TaskRef, period: Option<usize>) -> RealtimeTaskRef {
+        RealtimeTaskRef {
+            taskref: taskref,
+            period: period,
+            context_switches: 0,
+        }
+    }
+
+    /// Increment the number of times the task is picked
+    pub fn increment_context_switches(&mut self) {
+        self.context_switches = self.context_switches.saturating_add(1);
+    }
+
+    /// Checks whether the `RealtimeTaskRef` refers to a task that is periodic
+    pub fn is_periodic(&self) -> bool {
+        match self.period {
+            Some(_) => true,
+            None => false,
+        }
+    }
+
+    /// Checks whether the period of this `RealtimeTaskRef` is shorter than the period of another `RealtimeTaskRef`
+    /// If the `RealtimeTaskRef` is aperiodic, i.e. if `period` is `None`, we will always return false
+    /// Additionally, a periodic task will always return `true` if `other_taskref` is aperiodic
+    pub fn has_smaller_period(&self, other_taskref: &RealtimeTaskRef) -> bool{
+        match self.period {
+            Some(period_val) => if let Some(other_period_val) = other_taskref.period {
+                period_val < other_period_val
+            } else {
+                true
+            },
+            None => false,
+        }
+    }
+}
+
+lazy_static! {
+    /// There is one runqueue per core, each core only accesses its own private runqueue
+    /// and allows the scheduler to select a task from that runqueue to schedule in
+    static ref RUNQUEUES: AtomicMap<u8, RwLockIrqSafe<RunQueue>> = AtomicMap::new();
+}
+
+/// A list of references to `Task`s (`RealtimeTaskRef`s). 
+/// This is used to store the `Task`s (and associated scheduler related data) 
+/// that are runnable on a given core.
+/// A queue is used for the round robin scheduler.
+/// `Runqueue` implements `Deref` and `DerefMut` traits, which dereferences to `VecDeque`.
+#[derive(Debug)]
+pub struct RunQueue {
+    core: u8,
+    queue: VecDeque<RealtimeTaskRef>,
```

Comment:
I was thinking so myself, I was wondering whether you would know a good datatype for this use case, where we need:
1. Efficient insertion but also
2. The ability to iterate over the list. When we are selecting the next task, we want to select the task with the highest priority but with the condition that the task is currently Runnable. I don't think the BinaryHeap in alloc::collections would work as the into_iter() method does not iterate over the elements in sorted order, it only allows us to peek the min/max element.

I was thinking about using the priority queue implementation [here](https://docs.rs/priority-queue/latest/priority_queue/), but I tried using it as the priority queue implementation of the `sleep` crate but for some reason it pulls in `std` even when I set default-features = false in Cargo.toml?

If I can find the proper data type, I think I could implement this pretty quickly tonight

Reply:
Okay, I didn't realize you needed queue-like functionality.

That `priority-queue` crate does work for me, so you could certainly use it. The thing is, because we're in a `no_std` environment, you need to provide your own implementation of a hasher (which is normally provided by the std lib). Thus, you can't use the default constructors like `PriorityQueue::new()` or `with_capacity()`, you have to use the more specific constructors like this:

```rust
let mut q: PriorityQueue<&'static str, usize, ahash::RandomState> = priority_queue::PriorityQueue::with_default_hasher();
let mut q2: PriorityQueue<usize, usize, _> = PriorityQueue::with_hasher(ahash::RandomState::default());
```

where the `ahash` crate is providing the hasher implementation. So in the Cargo.toml file:
```toml
[dependencies]
priority-queue = "1.2.1"
ahash = { version = "0.7.6", default-features = false }
```